cubi_tk/sodar/update_samplehseet.py

```diff
@@ -0,0 +1,628 @@
+        assay_tsv = assay_final.to_csv(
+            sep="\t", index=False, header=list(map(orig_col_name, assay_final.columns))
+        )
+        ret = sodar_api.upload_ISA_samplesheet(
+            (isa_data["investigation"][0], investigation),
+            (isa_data["study"][0], study_tsv),
+            (isa_data["assay"][0], assay_tsv),
+        )
+        return ret
+
+    def parse_sampledata_args(self, isa_names) -> OrderedDict[str, str]:
+        """Build a dict of Sample-data filed name mappings and check consistency of provided data."""
+
+        # Some samples are defined, either through ped or sample data
+        if not self.args.ped and not self.args.sample_data:
+            raise ValueError(
+                "No sample data provided. Please provide either a ped file (`--ped`) or sample data (`-s`)."
+            )
+
+        # Base field name mapping (also used by ped)
+        sample_field_mapping = sheet_default_config[self.args.defaults]["field_column_mapping"]
+
+        # Sample data, if given, need to matche the defined fields (from default or sample_fields)
+        if self.args.sample_data:
+            if self.args.sample_fields:
+                sample_field_mapping.update(
+                    OrderedDict(
+                        [
+                            (field.split("=")[0], field.split("=")[1 if "=" in field else 0])
+                            for field in self.args.sample_fields
+                        ]
+                    )
+                )
+                n_fields = len(self.args.sample_fields)
+            else:
+                n_fields = len(sheet_default_config[self.args.defaults]["sample_fields"])
+
+            incorrect_n = [sample for sample in self.args.sample_data if len(sample) != n_fields]
+            if incorrect_n:
+                msg = (
+                    f"The number of entries for some samples does not match the number of fields defined ({n_fields}):\n"
+                    "\n".join(
+                        [
+                            f"{len(sample)} instead of {n_fields} values: {', '.join(map(str, sample))}"
+                            for sample in incorrect_n
+                        ]
+                    )
+                )
+                raise ValueError(msg)
+
+            unknown_fields = [
+                tup for tup in sample_field_mapping.items() if tup[1] not in isa_names.keys()
+            ]
+            if unknown_fields:
+                msg = (
+                    "Some columns for sample field mapping are not defined in the ISA samplesheet: "
+                    + ", ".join([f"{col} (from {field})" for field, col in unknown_fields])
+                )
+                raise NameError(msg)
+
+        return sample_field_mapping
+
+    def gather_ISA_column_names(
+        self, study: pd.DataFrame, assay: pd.DataFrame
+    ) -> dict[str, list[tuple[str, str]]]:
+        isa_regex = re.compile(r"(Characteristics|Parameter Value|Comment)\[(.*?)]")
```

Comment:
Is the `[` escaped or do we match a literal `\` here?

Reply:
It's escaped (pycharm complained if I also espaced the correspondign closing bracket)

Reply:
To get this straight, what is it supposed to match?

Reply:
It (is supposed to) match the column name that is shown in Sodar, but in ISA is encoded as a specific column type.
So it looks something like this in the ISAtab file:
`Characteristics[Column name]`
Having users always give that full name would be a bit much, I think